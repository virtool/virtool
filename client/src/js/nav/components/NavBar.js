import { MenuButton } from "@reach/menu-button";
import React from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { logout } from "../../account/actions";
import { Dropdown, DropdownMenuItem, DropdownMenuLink, DropdownMenuList, Icon, VTLogo } from "../../base";
import { getSoftwareUpdates } from "../../updates/actions";
import { NavBarItem } from "./NavBarItem";
import Update from "./Update";

const NavBarLeft = styled.div`
    display: flex;
    align-items: center;
`;

const NavBarRight = styled.div`
    display: flex;
    align-items: center;
    margin-right: calc(100% - 100vw + 20px);
`;

const NavBarLogo = styled(VTLogo)`
    margin: 0 30px 0 35px;
`;

const NavDropdownButton = styled(MenuButton)`
    align-items: center;
    background: transparent;
    border: none;
    color: ${props => props.theme.color.white};
    cursor: pointer;
    display: flex;
    height: 45px;
    outline: none;
    padding: 0 10px;

    :focus {
        color: ${props => props.theme.color.primaryDarkest};
    }

    *:not(:last-child) {
        margin-right: 3px;
    }
`;

const StyledNavBar = styled.div`
    background-color: ${props => props.theme.color.primary};
    color: white;
    display: flex;
    height: 45px;
    justify-content: space-between;
    position: fixed;
    top: 0;
    width: 100%;
    z-index: 90;
`;

export class Bar extends React.Component {
    componentDidMount() {
        this.props.onGet();
    }

    render() {
        return (
            <StyledNavBar>
                <NavBarLeft>
                    <NavBarLogo />
                    <NavBarItem to="/home">Home</NavBarItem>
                    <NavBarItem to="/jobs">Jobs</NavBarItem>
                    <NavBarItem to="/samples">Samples</NavBarItem>
                    <NavBarItem to="/refs">References</NavBarItem>
                    <NavBarItem to="/hmm">HMM</NavBarItem>
                    <NavBarItem to="/subtraction">Subtraction</NavBarItem>
                </NavBarLeft>

                <NavBarRight>
                    <Update />

                    <NavBarItem target="_blank" to="//gitter.im/virtool/virtool" rel="noopener noreferrer">
                        <Icon name="comments" />
                    </NavBarItem>

                    <NavBarItem target="_blank" to="//virtool.ca/docs/manual" rel="noopener noreferrer">
                        <Icon name="book" />
                    </NavBarItem>

                    {this.props.dev && (
                        <NavBarItem to={{ state: { devCommands: true } }}>
                            <Icon color="red" name="bug" />
                        </NavBarItem>
                    )}

                    <Dropdown>
                        <NavDropdownButton>
                            <Icon name="user" />
                            <span>{this.props.userId}</span>
                            <Icon name="caret-down" />
                        </NavDropdownButton>
                        <DropdownMenuList>
                            <DropdownMenuLink to="/account">Account</DropdownMenuLink>
                            {this.props.administrator && (
                                <DropdownMenuLink to="/administration">Administration </DropdownMenuLink>
                            )}
                            <DropdownMenuLink
                                to="//gitreports.com/issue/virtool/virtool"
                                target="_blank"
                                rel="noopener noreferrer"
                            >
                                Report Issue
                            </DropdownMenuLink>
                            <DropdownMenuItem onSelect={this.props.logout}>Logout</DropdownMenuItem>
                        </DropdownMenuList>
                    </Dropdown>
                </NavBarRight>
            </StyledNavBar>
        );
    }
}

export const mapStateToProps = state => ({
    ...state.account,
    dev: state.app.dev,
    pending: state.app.pending
});

export const mapDispatchToProps = dispatch => ({
    logout: () => {
        dispatch(logout());
    },

    onGet: () => {
        dispatch(getSoftwareUpdates());
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(Bar);
