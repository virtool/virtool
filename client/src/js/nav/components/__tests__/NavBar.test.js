import { Bar, mapStateToProps, mapDispatchToProps } from "../NavBar";

describe("<Bar />", () => {
    const props = {
        administrator: true,
        id: "foo",
        pending: false,
        onLogout: jest.fn()
    };
    it("should render", () => {
        const wrapper = shallow(<Bar {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("mapStateToProps", () => {
    const state = {
        account: { foo: "bar" },
        app: {
            pending: false
        }
    };
    it("should return props", () => {
        const result = mapStateToProps(state);
        expect(result).toEqual({
            foo: "bar",
            pending: false
        });
    });
});

describe("mapDispatchToProps", () => {
    const dispatch = jest.fn();

    it("should return onLogout in props", () => {
        const props = mapDispatchToProps(dispatch);
        props.onLogout();
        expect(dispatch).toHaveBeenCalled();
    });
});
