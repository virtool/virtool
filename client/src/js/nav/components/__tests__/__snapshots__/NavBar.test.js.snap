// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bar /> should render 1`] = `
<NavBar__StyledNavBar>
  <NavBar__NavBarLeft>
    <NavBar__NavBarLogo />
    <NavBarItem
      to="/home"
    >
      Home
    </NavBarItem>
    <NavBarItem
      to="/jobs"
    >
      Jobs
    </NavBarItem>
    <NavBarItem
      to="/samples"
    >
      Samples
    </NavBarItem>
    <NavBarItem
      to="/refs"
    >
      References
    </NavBarItem>
    <NavBarItem
      to="/hmm"
    >
      HMM
    </NavBarItem>
    <NavBarItem
      to="/subtraction"
    >
      Subtraction
    </NavBarItem>
  </NavBar__NavBarLeft>
  <NavBar__NavBarRight>
    <Connect(NotificationIcon) />
    <NavBarItem
      rel="noopener noreferrer"
      target="_blank"
      to="//gitter.im/virtool/virtool"
    >
      <Icon
        faStyle="fas"
        fixedWidth={false}
        name="comments"
      />
    </NavBarItem>
    <NavBarItem
      rel="noopener noreferrer"
      target="_blank"
      to="//virtool.ca/docs/manual"
    >
      <Icon
        faStyle="fas"
        fixedWidth={false}
        name="book"
      />
    </NavBarItem>
    <Menu>
      <NavBar__NavDropdownButton>
        <Icon
          faStyle="fas"
          fixedWidth={false}
          name="user"
        />
        <span />
        <Icon
          faStyle="fas"
          fixedWidth={false}
          name="caret-down"
        />
      </NavBar__NavDropdownButton>
      <Dropdown__DropdownMenuList>
        <Dropdown__DropdownMenuLink
          to="/account"
        >
          Account
        </Dropdown__DropdownMenuLink>
        <Dropdown__DropdownMenuLink
          to="/administration"
        >
          Administration 
        </Dropdown__DropdownMenuLink>
        <Dropdown__DropdownMenuLink
          rel="noopener noreferrer"
          target="_blank"
          to="//gitreports.com/issue/virtool/virtool"
        >
          Report Issue
        </Dropdown__DropdownMenuLink>
        <Dropdown__DropdownMenuItem
          onSelect={[MockFunction]}
        >
          Logout
        </Dropdown__DropdownMenuItem>
      </Dropdown__DropdownMenuList>
    </Menu>
  </NavBar__NavBarRight>
</NavBar__StyledNavBar>
`;
